website/blog/2022-01-31-avoid-data-silos-in-presto-in-meta.md

```diff
@@ -0,0 +1,171 @@
+Meta eventually replaced  it in favor of a new architecture based on local caching, namely RaptorX.
+
+![Raptor Timeline](/img/blog/2022-01-31-avoid-data-silos-in-presto-in-meta/timeline.png)
+
+## The story of Raptor
+
+Generally speaking, Presto as a query engine does not own storage. Instead, connectors were developed to query different external data sources.
+This framework is very flexible, but in disaggregated compute and storage architectures, it makes low latency guarantees hard to achieve due to the
+network / storage latency. To address this limitation, Raptor is designed as a shared-nothing storage engine for Presto.
+
+### Motivation – an initial use case in the AB testing framework
+
+<!--truncate-->
+
+In Meta, new product features typically  go through AB testing before they are released more broadly. The AB testing framework allows engineers to configure
+experiments by rolling out the feature to a test group, and monitors key metrics against a control group. The framework provides engineers
+with a UI to analyze their experiment’s statistics, which converts the configurations to Presto queries. The query shapes are known and limited.
+Queries typically join multiple large data sets, which include user, device, test, and event attributes, etc. The basic requirements for this use case are:
+1. **Losslessness**: data need to be complete and accurate
+2. **Flexibility**: users should be able to perform arbitrary slice and dice on their results
+3. **Freshness**: test results should be available within hours
+4. **Interactive latency**: Query needs to return results within seconds
+5. **High availability**: As a critical service for product development, there should be minimal downtime for the service.
+
+Presto in a typical warehouse setting (i.e., using Hive connector to query warehouse data directly) could easily meet the first two requirements but not
+the rest. At that time there was no near-real-time data ingestion and most warehouse data were ingested daily, thus not satisfying the freshness requirement.
+Meta's data center was already moving to disaggregated compute / storage architecture, thus hard to guarantee latency when scanning large tables with high
+QPS. Typical Presto deployment would stop the whole cluster, thus not satisfying HA requirements.
+
+To support this critical use case, we began the journey of productionizing Raptor.
+
+### Raptor Architecture
+Following is the high level architecture of a Presto cluster with Raptor connector.
+
+![Raptor Architecture](/img/blog/2022-01-31-avoid-data-silos-in-presto-in-meta/raptor_architecture.png)
+
+Raptor connector uses MySQL as its metastore, storing table and file metadata. Table data is stored on flash disks on each worker node and periodically
+backed up to an external storage system to enable recovery in case of worker node crash. Data is ingested into the Raptor cluster in mini batches with
```

Comment:
```suggestion
backed up to an external storage system to enable recovery in case of a worker node crashing. Data is ingested into the Raptor cluster in small enough batches to provide
```